### PR TITLE
Addresses Myriad 188 - NodeManager switch to UNHEALTHY causes NPE on …

### PR DESCRIPTION
**Pull request by DarinJ**
…ResourceManager. @sdaingade I believe this also addresses Myriad-156, which was assigned to you.  @kensipe can you look over the unit test mods?  I don't really know spock or groovy.  I'd like to add a test case for if the node doesn't exist in the yarnscheduler, though that may be difficult with mocks.

### CODE REVIEW

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java [hunk 1]
```diff
@@ -196,13 +196,22 @@ protected void handleContainerAllocation(RMNode rmNode) {
...
+      rmNode.getTotalCapability().setVirtualCores(newCapacity.getVirtualCores());
+      LOGGER.debug("Setting capacity for node {} to {}", rmNode.getHostName(), newCapacity);
+      // updates the scheduler with the new capacity for the NM.
+      // the event is handled by the scheduler asynchronously
+      synchronized (yarnScheduler) {
+        if (yarnScheduler.getSchedulerNode(rmNode.getNodeID()) != null) {
+          yarnScheduler.updateNodeResource(rmNode,
+              ResourceOption.newInstance(rmNode.getTotalCapability(), RMNode.OVER_COMMIT_TIMEOUT_MILLIS_DEFAULT));
+        } else {
+          LOGGER.info("Yarn Scheduler doesn't have node {}, probably UNHEALTHY", rmNode.getNodeID());
```
Comment (ZJaffee) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: I'm not entirely sure why you decided to go with the updateNodeResource method rather than scheduling a new NodeResourceUpdateSchedulerEvent, but from the looks of it they do relatively the same thing in theory.

That said, the comments look wrong, as now the event is handled by the scheduler synchronously, and the javadoc needs to be updated. Additionally the NodeResourceUpdateSchedulerEvent no longer needs to be imported.

File: myriad-scheduler/src/test/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy [hunk 2]
```diff
@@ -115,7 +117,7 @@ class YarnNodeCapacityManagerSpec extends FGSTestBaseSpec {
         then:
         zeroNM.getTotalCapability().getMemory() == 2048
         zeroNM.getTotalCapability().getVirtualCores() == 2
-        1 * rmContext.getDispatcher().getEventHandler().handle(_ as NodeResourceUpdateSchedulerEvent)
+        1 * yarnScheduler.updateNodeResource( _ as RMNode, _ as ResourceOption)
```
Comment (ZJaffee) [inline myriad-scheduler/src/test/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy]: I'm not confident that we can guarantee that this will run in a synchronized fashion, you might want to add the annotation above the "Set Node Capacity" function with something like ```@Synchronized("yarnScheduler")```

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java (hunk 1 above)
Comment (DarinJ) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: The reason was because NodeResourceUpdateSchedulerEvent eventually just calls `yarnSchedulers.updateResource`, this way avoids a possible race condition as we have a read/update issue.  Potentially a better way to fix this is to do the null pointer check in `updateNodeResource` and I may look into that for future versions of YARN.

File: myriad-scheduler/src/test/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy (hunk 2 above)
Comment (DarinJ) [inline myriad-scheduler/src/test/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy]: Is this only a concern about the test?  If so it's a mock interface, I'm not worried about the synchronization behavior here.  I just want to ensure the mocked method got called once.

Comment (ZJaffee) [inline myriad-scheduler/src/test/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy]: The reason I bring it up is that I believe that unless this test is synchronized we won't be testing the exact behavior of the method. Namely, the way the synchronized method is implemented above places a lock on the yarnScheduler object preventing other methods from being called on the object anywhere else in the class. Since by default groovy tests run in an asynchronous manner the behavior of the test may be different if the object is used anywhere else within the test class.

Comment (DarinJ) [inline myriad-scheduler/src/test/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManagerSpec.groovy]: I get that, but a mock is never going to handle that behavior and the method is a No Op anyway.  I agree we need better unit test coverage, but that would be a separate PR I think.

**Comment by DarinJ** [comment]
I've been running this for a few days now, while intermittently forcing nodes into unhealthy states (removing local dirs, changing permissions, etc).  It's clear from the logging it's working as expected, I'll merge tomorrow afternoon if I don't get any additional comments.
